syntax = "proto3";

// Package definition
package messaging;

// Enum for distinguishing the type of requests
enum RequestType {
  REQUEST_DMX = 0;
  REQUEST_POWER = 1;
  REQUEST_TELEMETRY = 2;
  REQUEST_CONFIGURATION = 3;
}

message DmxRequest {
  bytes values = 1;  // Single sequence of DMX values, up to 512 values
  bool ack = 2;  // If true, an acknowledgement from the device is requested
}

// These are settings that it's ok for the star to forget if it reboots.
message ConfigAmnesia {
  uint32 fogOutputOnMinTime = 1; // When the fog outputs, how many ms will it output?
  uint32 fogOutputOnMaxTime = 2; // When the fog outputs, how many ms will it output?
  uint32 fogOutputOffMinTime = 3; // For the random output, what's the minumim value?
  uint32 fogOutputOffMaxTime = 4; // For the random output, what's the minumim value?
  uint32 fogActivateTime = 5; // Turn on the fog for this duration.
  bool fogEnabled = 6; // Is the fog enabled?
}

// Enum for power devices
enum PowerDevices {
  FOG_MACHINE_A = 0;
  // Add other power devices as needed
}

// Enum for power request query
enum PowerQuery {
  UNDEFINED = 0;  // Query type is undefined
  CURRENTLY_USED = 1;  // Query for currently used power devices
  // Add other query types as needed
}

message PowerRequest {
  repeated PowerDevices devices = 1;  // Instruct the responder to use these powered devices and no other
  PowerQuery query = 2;  // The type of information about power devices requested
}

// Enum for telemetry request query
enum TelemetryQuery {
  TELEMETRY_ALL = 0;
  TELEMETRY_TEMPERATURE = 1;
  TELEMETRY_HUMIDITY = 2;
  TELEMETRY_PEAK_BOOM = 3;
  TELEMETRY_SYSTEM = 4;
}

message TelemetryRequest {
  TelemetryQuery query = 1;  // Type of telemetry information requested
}

// Enum for configuration request query
enum ConfigurationQuery {
  CONFIG_ALL = 0;
  CONFIG_BLOWER = 1;
}

message ConfigurationRequest {
  ConfigurationQuery query = 1;  // Type of configuration information requested
}

message Request {
  RequestType type = 1;  // Type of the request
  oneof request_payload {  // Request payload, only one of the fields can be set
    DmxRequest dmx_request = 2;
    PowerRequest power_request = 3;
    TelemetryRequest telemetry_request = 4;
    ConfigurationRequest configuration_request = 5;
  };
  ConfigAmnesia configAmnesia = 10;
}

// Enum for distinguishing the type of responses
enum ResponseType {
  RESPONSE_DMX = 0;
  RESPONSE_POWER = 1;
  RESPONSE_TELEMETRY = 2;
  RESPONSE_CONFIGURATION = 3;
  RESPONSE_ERROR = 10;  // Added new response type
}

// Enum for Response status
enum ResponseStatus {
  SUCCESS = 0;
  ERROR = 1;
}

message DmxResponse {
  ResponseStatus status = 1;  // This will only be returned if the request would like an acknowledgement
}

message PowerResponse {
  repeated PowerDevices power_used = 1;  // Devices that used power
  repeated PowerDevices power_requested = 2;  // Devices that requested power
}

message TelemetryResponse {
  float temperature = 1;  // Device temperature
  float humidity = 2;  // Device humidity
  float accel_min_z = 3;  // Z-axis acceleration
  float accel_max_z = 4;  // Z-axis acceleration
  float accel_min_z_time = 5;
  float accel_max_z_time = 6;
  string chip_model = 7;  // Device chip model
  string chip_revision = 8;  // Device chip revision
  uint32 chip_free_heap = 9;  // Device chip free heap
  string sdk_version = 10;  // Device SDK version
  string sketch_md5 = 11;  // Device sketch MD5
  string sketch_compile_datetime = 12;  // Device sketch compile datetime
  uint32 sketch_size = 13;  // Device sketch size
}

// Enum for device types in ConfigurationResponse
enum ConfigurationDeviceType {
  STAR = 0;
  // Add other device types as needed
}

message ConfigurationResponse {
  ConfigurationDeviceType device_type = 1;  // Type of the device
  uint32 blower = 2;  // 8-bit value for blower
}

// Added new ErrorResponse
message ErrorResponse {
  string message = 1;  // Description of the error
}

message Response {
  ResponseType type = 1;  // Type of the response
  oneof response_payload {  // Response payload, only one of the fields can be set
    DmxResponse dmx_response = 2;
    PowerResponse power_response = 3;
    TelemetryResponse telemetry_response = 4;
    ConfigurationResponse configuration_response = 5;
    ErrorResponse error_response = 6;  // Added ErrorResponse to the response payload
  }
}